File: src/inventory/service.ts
import prisma from "../lib/db.js";
import { AppError } from "../lib/middleware.js";

export async function get(productId: number) {
  const inventory = await prisma.inventory.findUnique({
    where: { productId },
  });
  if (!inventory) {
    throw new AppError(404, "Inventory not found");
  }
  return inventory;
}

export async function updateQuantity(productId: number, quantity: number) {
  const inventory = await prisma.inventory.findUnique({
    where: { productId },
  });
  if (!inventory) {
    throw new AppError(404, "Inventory not found");
  }

  return prisma.inventory.update({
    where: { productId },
    data: {
      quantity,
      lastUpdated: new Date(),
    },
  });
}

export async function reserve(productId: number, quantity: number) {
  const inventory = await prisma.inventory.findUnique({
    where: { productId },
  });
  if (!inventory) {
    throw new AppError(404, "Inventory not found");
  }

  const available = inventory.quantity - inventory.reserved;
  if (available < quantity) {
    throw new AppError(
      400,
      `Insufficient inventory. Available: ${available}, Requested: ${quantity}`,
    );
  }

  // Use atomic increment for better concurrency
  return prisma.inventory.update({
    where: { productId },
    data: {
      reserved: { increment: quantity },
      lastUpdated: new Date(),
    },
  });
}

File: src/ecommerce/inventory/service.py
"""Inventory business logic."""

from datetime import datetime

from fastapi import HTTPException
from sqlmodel.ext.asyncio.session import AsyncSession

from ecommerce.inventory.models import Inventory


async def get_inventory(session: AsyncSession, product_id: int) -> Inventory:
    """Get inventory for a product."""
    inventory = await session.get(Inventory, product_id)
    if not inventory:
        raise HTTPException(status_code=404, detail="Inventory not found")
    return inventory


async def update_inventory(
    session: AsyncSession, product_id: int, quantity: int
) -> Inventory:
    """Update inventory quantity."""
    inventory = await session.get(Inventory, product_id)
    if not inventory:
        raise HTTPException(status_code=404, detail="Inventory not found")

    inventory.quantity = quantity
    inventory.last_updated = datetime.utcnow()
    session.add(inventory)
    await session.commit()
    await session.refresh(inventory)
    return inventory


async def reserve_inventory(
    session: AsyncSession, product_id: int, quantity: int
) -> Inventory:
    """Reserve inventory for an order, checking availability."""
    inventory = await session.get(Inventory, product_id)
    if not inventory:
        raise HTTPException(status_code=404, detail="Inventory not found")

    available = inventory.quantity - inventory.reserved
    if available < quantity:
        raise HTTPException(
            status_code=400,
            detail=f"Insufficient inventory. Available: {available}, Requested: {quantity}",
        )

    inventory.reserved += quantity
    inventory.last_updated = datetime.utcnow()
    session.add(inventory)
    await session.commit()
    await session.refresh(inventory)
    return inventory
